doschema/jsvalidation.py

```diff
@@ -0,0 +1,320 @@
+
+        self._validate_root(schema, ())
+
+        root = self.fields_types_dict.get(())
+
+        if not root or root.field_type != 'object':
+            err_msg = "Root field / type is not 'object' in {0}"
+            raise JSONSchemaCompatibilityError(
+                '/',
+                self.uri,
+                err_msg.format(self.uri)
+            )
+
+        for path, field in six.iteritems(self.fields_types_dict):
+            if field is None:
+                logging.warning(
+                    'No type in field ' + self._make_json_pointer(path)
+                )
+
+        self.uri = None
+        self.resolver = None
+
+    def _validate_root(self, curr_schema, curr_field):
+        """Go through the schema and retrieve schema's particular fields.
+
+        :param curr_schema: Schema or subschema that is currently processed.
+        :param curr_field: Tuple with path to currently processed field.
+        """
+        while '$ref' in curr_schema:
+            path = curr_schema.get('$ref')
+            curr_schema = self.resolver.resolve(path)[1]
+
+        self._validate_collection(curr_schema, curr_field)
+
+        field = self.fields_types_dict.get(curr_field)
+
+        if "type" in curr_schema:
+            if ((curr_field in self.fields_types_dict and field is None) or
+                    (field is not None and field.field_type == 'null')):
+                raise NotImplementedError(
+                    'JSON schema null type is not supported.'
+                    )
+            elif field and curr_schema['type'] != field.field_type:
+                err_msg = "{0} type mismatch in schemas {1} and {2}."
+                raise JSONSchemaCompatibilityError(
+                    field.schema_index,
+                    self.uri,
+                    err_msg.format(
+                        self._make_json_pointer(curr_field),
+                        field.schema_index,
+                        self.uri
+                    )
+                )
+
+            field_type = curr_schema['type']
+
+            self.fields_types_dict[curr_field] = field = FieldToAdd(
+                schema_index=self.uri,
+                field_tuple=curr_field,
+                field_type=field_type
+            )
+
+            if curr_schema['type'] == 'array':
+                self._validate_array(
+                    curr_schema['items'],
+                    curr_field + ('items',)
+                )
+        if "enum" in curr_schema:
+            enum_type = self._validate_enum(
+                curr_schema['enum'],
+                curr_field
+            )
+            if field is not None and field.field_type != enum_type:
+                err_msg = """Conflicting type for field {0} previously found
+                            in schema {1} enum with type {2},
+                            found next with type {3}."""
+                raise JSONSchemaCompatibilityError(
+                    self.uri,
```

Comment:
Bug after refactoring? self.uri is not the previous schema. The previous schema can be left as None.